File: cpp/20_包含min函数的栈.cpp
#include <iostream>
#include <vector>
#include <stack>
#include <stdio.h>
#include <sys/time.h>

#include "my_vector.h"

// 等长最小栈，存在数据冗余

class Solution {
private:
    std::stack<int> stack;
    std::stack<int> minStack;

public:
    void push(int value) {
        if (minStack.empty() || value <= minStack.top()) {
            minStack.push(value);
        } else {
            minStack.push(minStack.top());
        } 
        stack.push(value);
    }

    void pop() {
        minStack.pop();
        stack.pop();
    }

    int top() {
        return stack.top();
    }

    int min() {
        return minStack.top();
    }

    // 测试函数
    // push随机数据，中间获取最小值
    // pop随机数据，中间获取最小值
    void test(std::vector<int>& vec) {
        struct timeval start, end;
        std::vector<int> ret;
        
        printf("=====\n");
        gettimeofday(&start, nullptr);
        for (auto & i : vec) {
            push(i);
            ret.push_back(min());
        }
        for (auto & i : vec) {
            ret.push_back(min());
            pop();
        }
        gettimeofday(&end, nullptr);
        printf("times(us): %ld, result: ", end.tv_usec - start.tv_usec);
        printf_1d_vec(ret);
    }
};

int main(int argc, char* argv[])
{
    std::vector<int> vec = vector_CreatRandom(1, 20, 5);
    std::vector<int> vec2 = vector_CreatRandom(1, 1000, 100);

    Solution s;
    s.test(vec);
    s.test(vec2);

    return 0;
}

File: python/20_包含min函数的栈.py
#!/bin/bash python3
#-*- coding:utf-8 -*-

class Solution:
    """
    注意最小数必须由一个栈来维护，不能是一个数
    否则经过pop之后就无法保持最小数是正确的了
    也就是说需要保持两个栈的对应性
    """
    def __init__(self):
        self.stack = []
        self.minStack = []

    def push(self, node):
        self.stack.append(node)
        if self.minStack == [] or self.min() > node:
            self.minStack.append(node)
        else:
            tmp = self.min()
            self.minStack.append(tmp)

    def pop(self):
        if self.stack == [] or self.minStack == []:
            return None
        self.stack.pop()
        self.minStack.pop()

    def top(self):
        return self.stack[-1]

    def min(self):
        if self.stack == [] or self.minStack == []:
            return None
        return self.minStack[-1]


def main():
    s = Solution()

    s.push(1)
    s.push(6)
    s.push(5)
    s.pop()
    s.push(8)
    s.push(0)
    s.push(3)

    print(s.top())
    print(s.min())


if __name__ == "__main__":
    main()